crates/holochain/src/core/workflow/sys_validation_workflow.rs

```diff
@@ -43,170 +42,424 @@ mod validate_op_tests;
-    //      there will be more ops to process after this workflow run.
-    let start = (start_len >= NUM_CONCURRENT_OPS).then(std::time::Instant::now);
-    let saturated = start.is_some();
-    let cascade = workspace.full_cascade(network);
-
-    validate_ops_batch(
-        sorted_ops,
-        start,
-        {
-            let space = space.clone();
-            move |so| {
-                // Create an incoming ops sender for any dependencies we find
-                // that we are meant to be holding but aren't.
-                // If we are not holding them they will be added to our incoming ops.
-                let incoming_dht_ops_sender =
-                    IncomingDhtOpSender::new(space.clone(), sys_validation_trigger.clone());
-                let workspace = workspace.clone();
-                let cascade = cascade.clone();
-                async move {
-                    let (op, op_hash) = so.into_inner();
-                    let op_type = op.get_type();
-                    let action = op.action();
-
-                    let dependency = get_dependency(op_type, &action);
-                    let dna_def = DnaDefHashed::from_content_sync((*workspace.dna_def()).clone());
-
-                    let r =
-                        validate_op(&op, &dna_def, &cascade, Some(&incoming_dht_ops_sender)).await;
-                    r.map(|o| (op_hash, o, dependency))
+
+    // Forget what dependencies are currently in use
+    current_validation_dependencies.lock().clear_retained_deps();
+
+    if sorted_ops.is_empty() {
+        tracing::trace!(
+            "Skipping sys_validation_workflow because there are no ops to be validated"
+        );
+
+        // If there's nothing to validate then we can clear the dependencies and save some memory.
+        current_validation_dependencies.lock().purge_held_deps();
+
+        return Ok(OutcomeSummary::new());
+    }
+
+    let num_ops_to_validate = sorted_ops.len();
+    tracing::debug!("Sys validating {} ops", num_ops_to_validate);
+
+    let cascade = Arc::new(workspace.local_cascade());
+    let dna_def = DnaDefHashed::from_content_sync((*workspace.dna_def()).clone());
+
+    retrieve_previous_actions_for_ops(
+        current_validation_dependencies.clone(),
+        cascade.clone(),
+        sorted_ops.clone().into_iter(),
+    )
+    .await;
+
+    // Now drop all the dependencies that we didn't just try to access while searching the current set of ops to validate.
+    current_validation_dependencies.lock().purge_held_deps();
+
+    let mut validation_outcomes = Vec::with_capacity(sorted_ops.len());
+    for hashed_op in sorted_ops {
+        let (op, op_hash) = hashed_op.into_inner();
+        let op_type = op.get_type();
+        let action = op.action();
+
+        // This is an optimization to skip app validation and integration for ops that are
+        // rejected and don't have dependencies.
```

Comment:
I don't understand, if this is meant for ops that don't have dependencies, why is `get_dependency` called?